File: src/Cpp/09_Listy_wprowadzenie/Zad16.cpp
#include <algorithm>
#include <cassert>
#include <vector>

/*
Dla otrzymanej listy oraz liczby x, znajdź indeksy pierwszej pary 
elementów listy, których suma jest równa liczbie x. Jeśli w liście 
nie ma takiej pary, zwróć parę (-1, -1).
*/

std::vector<std::pair<int, int>> znajdzParyV1(std::vector<int> &lista,
                                              int suma) {

  std::vector<std::pair<int, int>> wynik;

  for (unsigned int i = 0; i < lista.size(); i++) {
    for (unsigned int j = i + 1; j < lista.size(); j++) {
      if (lista[i] + lista[j] == suma)
        wynik.emplace_back(lista[i], lista[j]);
    }
  }

  return wynik;
}

void test1() {
  std::vector<int> lista{0, 4, 5, 6, 2, 9, 2, 3};
  std::vector<std::pair<int, int>> wynik{{0, 5}, {2, 3}, {2, 3}};
  int suma = 5;

  assert(znajdzParyV1(lista, suma) == wynik);
}

int main() {

  test1();

  return 0;
}

File: src/haskell/09_listy_wprowadzenie/zad16.hs
{-
ZAD-16 — Indeksy pierwszej pary o sumie x

\**Poziom:** ★★☆
\**Tagi:** `hashmap`, `indeksy`, `2-sum`

### Treść

Wczytaj listę liczb całkowitych oraz liczbę `x`. Znajdź indeksy **pierwszej** pary `(i, j)` (z `i < j`) takiej, że `lista[i] + lista[j] == x`.

Jeśli nie istnieje taka para — wypisz `-1 -1`.

### Wejście

\* 1. linia: `N` (`N ≥ 2`)
\* kolejne `N` linii: liczby całkowite
\* ostatnia linia: `x` (liczba całkowita)

### Wyjście

Dwie liczby całkowite w jednej linii (oddzielone spacją): `i j` albo `-1 -1`.

### Przykład

\**Wejście:**

```
5
1
3
4
5
2
5
```

\**Wyjście:**

```
0 2
```

-}
main :: IO ()
main = do
  n <- readLn :: IO Int
  nums <- mapM (\_ -> readLn :: IO Int) [1..n]
  x <- readLn :: IO Int
  
  let findPair [] = Nothing
      findPair (a:as) = case [j | (j, b) <- zip [1..] as, a + b == x] of
        (j:_) -> Just (0, j)
        [] -> case findPair as of
          Just (i, j) -> Just (i+1, j+1)
          Nothing -> Nothing
  
  case findPair nums of
    Just (i, j) -> putStrLn $ show i ++ " " ++ show j
    Nothing -> putStrLn "-1 -1"
